braille-printer-app/brf-printer-app.c

```diff
@@ -0,0 +1,821 @@
+
+  return (best_name);
+}
+
+
+
+//
+// 'match_id()' - Compare two IEEE-1284 device IDs and return a score.
+//
+// The score is 2 for each exact match and 1 for a partial match in a comma-
+// delimited field.  Any non-match results in a score of 0.
+//
+
+static int				// O - Score
+match_id(int           num_did,		// I - Number of device ID key/value pairs
+         cups_option_t *did,		// I - Device ID key/value pairs
+         const char    *match_id)	// I - Driver's device ID match string
+{
+  int		i,			// Looping var
+		score = 0,		// Score
+		num_mid;		// Number of match ID key/value pairs
+  cups_option_t	*mid,			// Match ID key/value pairs
+		*current;		// Current key/value pair
+  const char	*value,			// Device ID value
+		*valptr;		// Pointer into value
+
+
+  // Parse the matching device ID into key/value pairs...
+  if ((num_mid = papplDeviceParseID(match_id, &mid)) == 0)
+    return (0);
+
+  // Loop through the match pairs to find matches (or not)
+  for (i = num_mid, current = mid; i > 0; i --, current ++)
+  {
+    if ((value = cupsGetOption(current->name, num_did, did)) == NULL)
+    {
+      // No match
+      score = 0;
+      break;
+    }
+
+    if (!strcasecmp(current->value, value))
+    {
+      // Full match!
+      score += 2;
+    }
+    else if ((valptr = strstr(value, current->value)) != NULL)
+    {
+      // Possible substring match, check
+      size_t mlen = strlen(current->value);
+					// Length of match value
+      if ((valptr == value || valptr[-1] == ',') && (!valptr[mlen] || valptr[mlen] == ','))
+      {
+        // Partial match!
+        score ++;
+      }
+      else
+      {
+        // No match
+        score = 0;
+        break;
+      }
+    }
+    else
+    {
+      // No match
+      score = 0;
+      break;
+    }
+  }
+}
+
+//
+// 'driver_cb()' - Main driver callback.
+//
+
+static bool				// O - `true` on success, `false` on error
+driver_cb(
+    pappl_system_t         *system,	// I - System
+    const char             *driver_name,// I - Driver name
+    const char             *device_uri,	// I - Device URI
+    const char             *device_id,	// I - 1284 device ID
+    pappl_pr_driver_data_t *data,	// I - Pointer to driver data
+    ipp_t                  **attrs,	// O - Pointer to driver attributes
+    void                   *cbdata)	// I - Callback data (not used)
+{
+  int	i;				// Looping var
+
+
+  // Copy make/model info...
+  for (i = 0; i < (int)(sizeof(brf_drivers) / sizeof(brf_drivers[0])); i ++)
+  {
+    if (!strcmp(driver_name, brf_drivers[i].name))
+    {
+      papplCopyString(data->make_and_model, brf_drivers[i].description, sizeof(data->make_and_model));
+      break;
+    }
+  }
+
+  // Pages per minute 
+  data->ppm = 60;
+
+ 
+  // Color values...
+  data->color_supported   = PAPPL_COLOR_MODE_AUTO | PAPPL_COLOR_MODE_MONOCHROME;
+  data->color_default   = PAPPL_COLOR_MODE_MONOCHROME;
+  data->raster_types  = PAPPL_PWG_RASTER_TYPE_BLACK_8; // to be done just guess
```

Comment:
I raster input really 8-bit grayscale? And not 1-bit bi-level? (Depends on the filters)